file/shard.go

```diff
@@ -77,9 +84,86 @@ func (s *shardNodeReader) makeReader() (io.Reader, error) {
 	if len(readers) == 0 {
 		return nil, io.EOF
 	}
+	s.len = at
 	return io.MultiReader(readers...), nil
 }
 
+func (s *shardNodeFile) unpack() (data.UnixFSData, error) {
+	var retErr error
+	s.unpackLk.Do(func() {
+		nodeData, err := s.substrate.LookupByString("Data")
+		if err != nil {
+			retErr = err
+			return
+		}
+		nodeDataBytes, err := nodeData.AsBytes()
+		if err != nil {
+			retErr = err
+			return
+		}
+		ud, err := data.DecodeUnixFSData(nodeDataBytes)
+		if err != nil {
+			retErr = err
+			return
+		}
+		s.metadata = ud
+	})
+	return s.metadata, retErr
+}
+
+// returns the size of the n'th link from this shard.
+// the io.ReadSeeker of the child will be return if it was loaded as part of the size calculation.
+func (s *shardNodeFile) linkSize(lnk ipld.Node, position int) (int64, io.ReadSeeker, error) {
+	lnkhash, err := lnk.LookupByString("Hash")
+	if err != nil {
+		return 0, nil, err
+	}
+	lnklnk, err := lnkhash.AsLink()
+	if err != nil {
+		return 0, nil, err
+	}
+	_, c, err := cid.CidFromBytes([]byte(lnklnk.Binary()))
+	if err != nil {
+		return 0, nil, err
+	}
+
+	// efficiency shortcut: for raw blocks, the size will match the bytes of content
+	if c.Prefix().Codec == cid.Raw {
+		size, err := lnk.LookupByString("Tsize")
+		if err != nil {
+			return 0, nil, err
+		}
+		sz, err := size.AsInt()
+		return sz, nil, err
+	}
+
+	// check if there are blocksizes written, use them if there are.
```

Comment:
```suggestion
	// check if there are blocksizes written, use them if there are.
	// both err and md can be nil if this was not the first time unpack()
	// was called but there was an error on the first call.
```

confused me for a moment why the complicated `if` below, maybe more notes will help others